[Record Function] Add Kwargs only USER_SCOPE Macro

sraikund16 opened the pull request:
Summary: Add a macro such that users can easily add a USER annotation with kwargs only

Test Plan: Will use D63801503 to test this E2E. Added unit test as well that makes sure that the kwargs get recorded correctly

Differential Revision: D64420328

sraikund16 commented on test/cpp/profiler/record_function.cpp:
```diff
@@ -303,3 +303,31 @@ TEST(RecordFunctionTest, MultipleCallbacks) {
   ASSERT_FALSE(at::hasCallbacks());
 }
+
+// Test that callbacks are run in the order that they are registered.
```
oops fix this comment

davidberard98 commented on aten/src/ATen/record_function.h:
```diff
@@ -299,6 +299,31 @@ struct TORCH_API RecordFunction {
+    before(std::move(fn), args, current_sequence_nr);
+  }
+
+  template <typename F>
+  void before(
+      F fn,
+      const std::unordered_map<std::string, IValue>* kwargs,
+      int64_t current_sequence_nr = -1) {
+    if (!isActive()) {
+      return;
+    }
+    kwinputs_ = *kwargs;
+    before(fn, current_sequence_nr);
+  }
```
is this variant needed? it seems like it's not used in the macro that you added

sraikund16 replied:
It is because in this line https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/record_function.h#L540
needsInputs() can be false but we have kwargs as the first variable in ##__VA_ARGS__ so we need to abide by a strict template that has kwargs and a sequence_num in that order. 

I was thinking since this template is only used in this case today, that we could just ignore kwargs and not bother storing it, but then thought that in the future the template might be used for other instances in which we want to save the kwargs.

davidberard98 reviewed (approved):
minor comment about record_function.h changes, otherwise lgtm!

sraikund16 committed:
[Record Function] Add Kwargs only USER_SCOPE Macro (#138020)

Summary:

Add a macro such that users can easily add a USER annotation with kwargs only

Test Plan: Will use D63801503 to test this E2E. Added unit test as well that makes sure that the kwargs get recorded correctly

Reviewed By: davidberard98

Differential Revision: D64420328
